### PR TITLE
[Frontend] Support constant negative step sizes in for loops

### LINKED ISSUE #50: [Frontend] Catalyst for loop does not support negative step size
As the Catalyst control flow maps directly to MLIR control flow from the `SCF` dialect, the following loop configuration is not supported for example: 

```py
@qjit
def revc():
    @for_loop(9, -1, -1)
    def loop(i, agg):
        return agg + 1
    return loop(0)

def revi():
    agg = 0
    for y in range(9, -1, -1):
        agg += 1
    return agg

assert revc.mlir
assert revc() == revi()
```

See the documentation in MLIR for more details of the structured control flow operations: https://mlir.llvm.org/docs/Dialects/SCFDialect/#scffor-mlirscfforop.

However, while negative bounds and/or stepsizes are not directly supported in MLIR, I think we could support them in the frontend via a transformation on the loop index during the lowering phase.

### PR DESCRIPTION
**Pull request by pengmai**
**Context:** `@for_loop`s do not support negative step sizes.

**Description of the Change:** When passed in a constant step size that is negative, the lowering pipeline will perform a transformation. To simulate `for (i = start; i > stop; i -= step)`:
1. The loop is normalized to iterate from 0 to the computed # of iterations.
2. The induction variable (`iv`) is transformed to recompute the correct `iv` from the normalized `iv` at each iteration.

**Benefits:** Loops are now able to support negative step sizes. When a loop stop point would never be reached (e.g. `for (i = 0; i < 10; i -= 1)` or `for (int i = 5; i > -5; i += 1)`, the loop iterates 0 times. This behavior is consistent with Python's `range`.

**Possible Drawbacks:** This only supports **constant** step sizes. Supporting non-constant step sizes would require detecting if the step size was negative at runtime, which would either increase code complexity (hurting readability of the IR), or add overhead for loops with positive step sizes at each iteration.

If a non-constant step size is passed to the loop that evaluates to a negative value, the loop will incorrectly iterate 0 times. We could consider adding a runtime assertion (or warning) to detect this case.

**Related GitHub Issues:** Closes #50

### CODE REVIEW

File: frontend/catalyst/jax_primitives.py [hunk 1]
```diff
@@ -1262,6 +1270,20 @@ def _qfor_lowering(
...
         loop_operands.append(p)
+
+    if apply_reverse_transform:
+        zero = ArithConstantOp(ir.IndexType.get(), 0)
+        one = ArithConstantOp(ir.IndexType.get(), 1)
+        start_val, stop_val, step_val = loop_operands[0], loop_operands[1], loop_operands[2]
+
+        # Iterate from 0 to the number of iterations (ceil((start - stop) / -step))
+        distance = SubIOp(start_val, stop_val)
+        negative_step = SubIOp(zero, step_val)
```
Comment (dime10) [inline frontend/catalyst/jax_primitives.py:1281]: Is `step_val` not already negative? Or I suppose this is to make it positive? 😅

**Review by dime10 — COMMENTED** [review]
Sweet, so simple!

The changelog and docstring for `catalyst.for_loop` would need updating :)

> If a non-constant step size is passed to the loop that evaluates to a negative value, the loop will incorrectly iterate 0 times. We could consider adding a runtime assertion (or warning) to detect this case.

This is a pre-existing issue right?

File: frontend/catalyst/jax_primitives.py (hunk 1 above)
Comment (dime10) [reply frontend/catalyst/jax_primitives.py:1281]: Oh right, I see the comment now. The variable name was confusing me but it does make sense as a negation of the step size (making it positive in this case) 😅

**Comment by pengmai** [comment]
> This is a pre-existing issue right?

Yup, this is consistent with the current behavior.

**Pull request merged**